pubnub-hyper/src/transport/hyper/pam.rs

```diff
@@ -0,0 +1,275 @@
+    let map = |input: &HashMap<String, pam::BitMask>| -> json::JsonValue {
+        let mut data = json::JsonValue::new_object();
+        for (key, val) in input {
+            data[key] = val.0.into();
+        }
+        data
+    };
+    let resources = {
+        let input = &input.permissions.resources;
+        let mut data = json::JsonValue::new_object();
+        data["channels"] = map(&input.channels);
+        data["groups"] = map(&input.groups);
+        data["users"] = map(&input.users);
+        data["spaces"] = map(&input.spaces);
+        data
+    };
+    let patterns = {
+        let input = &input.permissions.patterns;
+        let mut data = json::JsonValue::new_object();
+        data["channels"] = map(&input.channels);
+        data["groups"] = map(&input.groups);
+        data["users"] = map(&input.users);
+        data["spaces"] = map(&input.spaces);
+        data
+    };
+    let permissions = {
+        let mut data = json::JsonValue::new_object();
+        data["resources"] = resources;
+        data["patterns"] = patterns;
+        data["meta"] = input.permissions.meta;
+        data
+    };
+    let grant_body = {
+        let mut data = json::JsonValue::new_object();
+        data["ttl"] = input.ttl.into();
+        data["permissions"] = permissions;
+        data
+    };
+    json::stringify(grant_body)
+}
+
+/// Obtain UNIX timestamp.
+fn get_unix_time() -> u64 {
+    let current = std::time::SystemTime::now();
+    let since_the_epoch = current
+        .duration_since(std::time::UNIX_EPOCH)
+        .expect("things seem to be happening before the unix epoch, check system clock");
+    since_the_epoch.as_secs()
+}
+
+/// Prepare the signature.
+fn prepare_signature(
+    secret_key: &str,
+    subscribe_key: &str,
+    publish_key: &str,
+    timestamp: u64,
+    body: &str,
+) -> String {
+    pam_signature::sign(
+        secret_key,
+        pam_signature::Request {
+            publish_key,
+            method: "POST",
+            path: &format!("/v3/pam/{}/grant", subscribe_key),
+            query: &format!("timestamp={}", timestamp),
+            body,
+        },
+    )
+}
+
+async fn handle_grant_response(response: Response<Body>) -> Result<response::Grant, error::Error> {
+    match response.status() {
+        StatusCode::OK => {
+            let data_json = handle_json_response(response).await?;
+            let err_fn = || error::Error::UnexpectedResponseSchema(data_json.clone());
+            let token = {
+                let data = json_as_object(&data_json["data"]).ok_or_else(err_fn)?;
+                let token = data["token"].as_str().ok_or_else(err_fn)?;
+                token.to_owned()
+            };
+            Ok(token)
+        }
+        StatusCode::BAD_REQUEST | StatusCode::FORBIDDEN => {
+            let data = handle_json_response(response).await?;
+            let error_message: String = format!("{}", data["error"]["message"]);
```

Comment:
Makes sense! Yes, this is something I intentionally omitted for now. I was planing to add this at this PR, but then decided to add it separately, so that this PR doesn't grow too big.

I'd prefer to generalize the common error type that services return around the error response type. This will require some adjustments to the transport and service traits, but I think this is the way to go. But the main problem is it'll require GAT.
Noting that I was also thinking about adding another parametrization value for the calls to pass as an input, besides the request - to allow requesting and passing transport-specific metadata.

With that said, I'd rather address the design for error types separately.